src/pybind/mgr/rbd_support/module.py

```diff
@@ -483,13 +501,735 @@ def get_perf_counters(self, pool_spec, sort_by):
+
+            return Task(d[TASK_SEQUENCE], d[TASK_ID], d[TASK_MESSAGE], d[TASK_REFS])
+        except json.JSONDecodeError as e:
+            raise ValueError("Invalid JSON ({})".format(str(e)))
+        except KeyError as e:
+            raise ValueError("Invalid task format (missing key {})".format(str(e)))
+
+
+class TaskHandler:
+    lock = Lock()
+    condition = Condition(lock)
+    thread = None
+
+    in_progress_task = None
+    tasks_by_sequence = dict()
+    tasks_by_id = dict()
+
+    sequence = 0
+
+    def __init__(self, module):
+        self.module = module
+        self.log = module.log
+
+        with self.lock:
+            self.init_task_queue()
+
+        self.thread = Thread(target=self.run)
+        self.thread.start()
+
+    @property
+    def default_pool_name(self):
+        return self.module.get_ceph_option("rbd_default_pool")
+
+    def extract_pool_spec(self, pool_spec):
+        pool_spec = extract_pool_key(pool_spec)
+        if pool_spec == GLOBAL_POOL_KEY:
+            pool_spec = (self.default_pool_name, '')
+        return pool_spec
+
+    def extract_image_spec(self, image_spec):
+        match = re.match(r'^(?:([^/]+)/(?:([^/]+)/)?)?([^/@]+)$',
+                         image_spec or '')
+        if not match:
+            raise ValueError("Invalid image spec: {}".format(image_spec))
+        return (match[1] or self.default_pool_name, match[2] or '', match[3])
+
+    def run(self):
+        try:
+            self.log.info("TaskHandler: starting")
+            while True:
+                with self.lock:
+                    now = datetime.now()
+                    for sequence in sorted([sequence for sequence, task
+                                            in self.tasks_by_sequence.items()
+                                            if not task.retry_time or task.retry_time <= now]):
+                        self.execute_task(sequence)
+
+                    self.condition.wait(5)
+                    self.log.debug("TaskHandler: tick")
+
+        except Exception as ex:
+            self.log.fatal("Fatal runtime error: {}\n{}".format(
+                ex, traceback.format_exc()))
+
+    @contextmanager
+    def open_ioctx(self, spec):
+        try:
+            with self.module.rados.open_ioctx(spec[0]) as ioctx:
+                ioctx.set_namespace(spec[1])
+                yield ioctx
+        except rados.ObjectNotFound:
+            self.log.error("Failed to locate pool {}".format(spec[0]))
+
+    @classmethod
+    def format_image_spec(cls, image_spec):
+        image = image_spec[2]
+        if image_spec[1]:
+            image = "{}/{}".format(image_spec[1], image)
+        if image_spec[0]:
+            image = "{}/{}".format(image_spec[0], image)
+        return image
+
+    def init_task_queue(self):
+        for pool_id, pool_name in get_rbd_pools(self.module).items():
+            try:
+                with self.module.rados.open_ioctx2(pool_id) as ioctx:
```

Comment:
By some reason python2 does not think `pool_id` is int here:
```
2019-07-25T11:23:00.522+0300 7ff87e52d700 -1 mgr load Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 1207, in __init__
    self.task = TaskHandler(self)
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 607, in __init__
    self.init_task_queue()
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 668, in init_task_queue
    with self.module.rados.open_ioctx2(pool_id) as ioctx:
  File "rados.pyx", line 564, in rados.requires.wrapper.validate_func (/home/mgolub/ceph/ceph.ci/build/src/pybind/rados2/rados.c:8341)
  File "rados.pyx", line 552, in rados.requires.check_type (/home/mgolub/ceph/ceph.ci/build/src/pybind/rados2/rados.c:7732)
TypeError: pool_id must be int
```
`self.module.rados.open_ioctx2(int(pool_id))` fixes this.

Reply:
Fixed